File: java/euler/src/main/java/com/rbleggi/euler/P02.java
package com.rbleggi.euler;

import java.util.stream.Stream;

public class P02 {
    static class FibGenerator {
        private long a = 1;
        private long b = 2;
        private boolean first = true;
        private boolean second = true;

        public long next() {
            if (first) {
                first = false;
                return a;
            }
            if (second) {
                second = false;
                return b;
            }
            long next = a + b;
            a = b;
            b = next;
            return next;
        }
    }

    public static void main(String[] args) {
        var fibGen = new FibGenerator();
        long evenSum = Stream.generate(fibGen::next)
            .takeWhile(fib -> fib <= 4_000_000)
            .filter(fib -> fib % 2 == 0)
            .mapToLong(Long::longValue)
            .sum();
        System.out.println("Sum of even Fibonacci numbers not exceeding 4 million: " + evenSum);
    }
}

File: kotlin/euler/src/main/kotlin/com/rbleggi/euler/P02.kt
package com.rbleggi.euler

fun mainP02() {
    fun generateFibs() = sequence {
        var a = 1
        var b = 2
        yield(a)
        yield(b)
        while (true) {
            val next = a + b
            yield(next)
            a = b
            b = next
        }
    }

    val fibsUpTo4M = generateFibs().takeWhile { it <= 4000000 }
    val evenSum = fibsUpTo4M.filter { it % 2 == 0 }.sum()
    println("Sum of even Fibonacci numbers not exceeding 4 million: $evenSum")
}
